### PR TITLE
feat: add cross-gateway peer fan-out to all per-entity resource handlers

### LINKED ISSUE #371: [BUG] Per-entity resource handlers missing cross-gateway peer fan-out
# Bug report

### Steps to reproduce
1. Configure two gateways with aggregation (primary + secondary peer)
2. Define a cross-ECU function in manifests with hosts on both gateways
3. Query per-entity resource collections on the primary gateway:
   - `GET /functions/{id}/logs`
   - `GET /functions/{id}/data`
   - `GET /functions/{id}/operations`
   - `GET /functions/{id}/faults`
   - `GET /functions/{id}/configurations`

### Expected behavior

Per-entity resource collection endpoints should fan-out to peer gateways via `AggregationManager::fan_out_get` and merge peer results into the response. This is necessary for cross-ECU functions where hosted apps live on different gateways (e.g., ROS 2 apps on primary, RTMaps apps on secondary).

### Actual behavior

Only the global `GET /faults` endpoint fans out to peers. All per-entity resource endpoints (logs, data, operations, faults, configurations) query only the local manager. Peer apps without local FQNs (e.g., RTMaps instances) are invisible in aggregated responses.

For functions, the handler collects FQNs from hosted apps via `effective_fqn()`. Peer apps return empty FQNs (no local ROS binding), so they get skipped. The handler never queries the peer gateway that actually owns those apps.

### Environment
- ros2_medkit version: main (c5cfc979)
- ROS 2 distro: Jazzy
- OS: Ubuntu 24.04

### Additional information
Affects all 5 resource collection handlers: log_handlers, data_handlers, operation_handlers, fault_handlers (per-entity), config_handlers. The global `/faults` endpoint already has fan-out - the same pattern needs to be applied to per-entity endpoints.

### PR DESCRIPTION
**Pull request by bburda**
# Pull Request

## Summary

Per-entity resource collection endpoints (`/{entity_type}/{id}/logs`, `/data`, `/operations`, `/faults`, `/configurations`) now fan-out GET requests to peer gateways via `AggregationManager::fan_out_get` and merge peer results into the response. This enables cross-ECU functions where hosted apps live on different gateways.

Introduces a shared `fan_out_helpers.hpp` with reusable inline utilities (`url_encode_param`, `build_fan_out_path`, `merge_peer_items`) used by all 5 handlers.

---

## Issue

- closes #371

---

## Type

- [x] Bug fix
- [x] New feature or tests
- [ ] Breaking change
- [ ] Documentation only

---

## Testing

- All 3716 unit tests pass
- All 3668 lint tests pass (0 failures)
- New integration test `test_cross_ecu_fanout` with two hybrid-mode gateways in separate DDS domains, cross-ECU function `vehicle_health`, validates fan-out for all 5 resource types:
  - Logs: peer log entries merged, severity filter forwarded, x-medkit aggregation metadata
  - Data: peer topics merged into function data response
  - Operations: peer services (lidar calibrate) merged with local (engine calibrate)
  - Faults: peer lidar faults (LIDAR_RANGE_INVALID) visible on primary via fan-out
  - Configurations: peer app parameters merged into function config response
  - Peer isolation: secondary has no primary logs (DDS domain isolation verified)

---

## Checklist

- [x] Breaking changes are clearly described (and announced in docs / changelog if needed)
- [x] Tests were added or updated if needed
- [ ] Docs were updated if behavior or public API changed

### CODE REVIEW

File: src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+)
+
+
+PRIMARY_PORT = get_test_port(0)
+PEER_PORT = get_test_port(1)
+PRIMARY_URL = f'http://localhost:{PRIMARY_PORT}{API_BASE_PATH}'
+PEER_URL = f'http://localhost:{PEER_PORT}{API_BASE_PATH}'
+
+PRIMARY_DOMAIN_ID = get_test_domain_id(0)
+PEER_DOMAIN_ID = get_test_domain_id(1)
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py:65]: Fixed in e047abf3.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 2]
```diff
@@ -410,6 +411,7 @@ void FaultHandlers::handle_list_faults(const httplib::Request & req, httplib::Re
       }
       ext.add("aggregation_sources", source_ids);
 
+      merge_peer_items(ctx_.aggregation_manager(), req, response, ext);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp:413]: Fixed in 796e45b9 - count moved after merge_peer_items in all 4 entity type branches (FUNCTION, COMPONENT, AREA, APP).

File: src/ros2_medkit_gateway/src/http/handlers/data_handlers.cpp [hunk 3]
```diff
@@ -128,6 +129,7 @@ void DataHandlers::handle_list_data(const httplib::Request & req, httplib::Respo
       resp_ext.add("aggregation_sources", aggregated.source_ids);
       resp_ext.add("aggregation_level", aggregated.aggregation_level);
     }
+    merge_peer_items(ctx_.aggregation_manager(), req, response, resp_ext);
     response["x-medkit"] = resp_ext.build();
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/data_handlers.cpp:135]: Fixed in e047abf3 - total_count moved after merge_peer_items.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  if (agg == nullptr) {
+    return;
+  }
+  auto fan_path = build_fan_out_path(req);
+  auto fan_result = agg->fan_out_get(fan_path, req.get_header_value("Authorization"));
+  if (fan_result.merged_items.is_array()) {
+    for (const auto & item : fan_result.merged_items) {
+      result["items"].push_back(item);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:90]: Fixed in 796e45b9 - merge_peer_items now ensures result["items"] exists as array before appending.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  std::string path = req.path;
+  char sep = '?';
+  for (const auto & [key, value] : req.params) {
+    path += sep;
+    path += key;
+    path += '=';
+    path += url_encode_param(value);
+    sep = '&';
+  }
+  return path;
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:61]: Fixed in e047abf3 - both keys and values are now URL-encoded.

File: src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+            'Secondary',
+        )
+        cls._wait_for_apps(
+            PRIMARY_URL,
+            {
+                'temp_sensor', 'rpm_sensor', 'calibration',
+                'pressure_sensor', 'actuator', 'lidar_sensor',
+            },
+            'Primary (merged)',
+        )
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py:297]: GatewayTestCase is designed for single-gateway tests (single BASE_URL). Multi-gateway tests need two URLs (primary + peer) with independent polling, which is the same pattern used by test_peer_aggregation.test.py.

File: src/ros2_medkit_gateway/src/http/handlers/log_handlers.cpp [hunk 7]
```diff
@@ -69,86 +69,62 @@ void LogHandlers::handle_get_logs(const httplib::Request & req, httplib::Respons
...
+        }
+        ext.add("aggregation_sources", log_source_fqns);
+      }
     }
-    ext.add("aggregation_sources", log_source_fqns);
-    result["x-medkit"] = ext.build();
 
+    merge_peer_items(ctx_.aggregation_manager(), req, result, ext);
+    result["x-medkit"] = ext.build();
     HandlerContext::send_json(res, result);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/log_handlers.cpp:118]: Design doc, config doc, and tutorial updated in e047abf3 to describe fan-out for all resource collections. REST API reference (rest.rst) update deferred to follow-up.

File: src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+    # @verifies REQ_INTEROP_061
+    def test_logs_severity_filter_forwarded(self):
+        """Severity filter is forwarded to peer via query params."""
+        data = _get_json(
+            f'{PRIMARY_URL}{FUNC_ENDPOINT}/logs?severity=fatal'
+        )
+        self.assertIsNotNone(data)
+        for entry in data.get('items', []):
+            self.assertEqual(entry['severity'], 'fatal')
+
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py:409]: Acknowledged - error-case tests (peer-down scenarios, invalid entity fan-out) will be added as follow-up.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  if (fan_result.merged_items.is_array()) {
+    for (const auto & item : fan_result.merged_items) {
+      result["items"].push_back(item);
+    }
+  }
+  if (fan_result.is_partial) {
+    ext.add("partial", true);
+    ext.add("failed_peers", fan_result.failed_peers);
+  }
+}
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:95]: Acknowledged - unit tests for fan_out_helpers.hpp (url_encode_param edge cases, build_fan_out_path, merge_peer_items with mock AggregationManager) will be added as follow-up.

[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:95]: Added in c1cca528 - test_fan_out_helpers.cpp with 13 unit tests covering url_encode_param (empty, unreserved, special chars, unicode), build_fan_out_path (no params, single/multiple, key encoding), and merge_peer_items (null agg, no-fan-out header).

File: src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py:409]: Added in c1cca528 - error-case integration tests: nonexistent function returns 404 (validated before fan-out), invalid severity returns 400. Also added DDS isolation assertion for data topics.

File: docs/config/aggregation.rst [hunk 10]
```diff
@@ -366,16 +366,25 @@ during periodic cache refresh cycles. These endpoints do not perform real-time
...
+Per-entity resource collection endpoints perform real-time fan-out via
+``fan_out_get()`` to collect items from all healthy peers and merge them
+into the local response. This applies to:
+
+- ``GET /{entity_type}/{id}/data``
+- ``GET /{entity_type}/{id}/operations``
+- ``GET /{entity_type}/{id}/faults``
+- ``GET /{entity_type}/{id}/configurations``
+- ``GET /{entity_type}/{id}/logs``
+- ``GET /faults`` (global fault list)
```
[the remark this answers is not in the document]
Comment (bburda) [reply docs/config/aggregation.rst]: Fixed in 72f34308 - added /api/v1/ prefix to all endpoint examples.

File: src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+  req.path = "/api/v1/test";
+  json result;
+  // No "items" key at all
+  XMedkit ext;
+
+  merge_peer_items(nullptr, req, result, ext);
+
+  // null agg = no-op, items should not be created
+  EXPECT_FALSE(result.contains("items"));
+}
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp]: Fixed in 72f34308 - renamed to merge_peer_items_null_agg_does_not_touch_result.

File: src/ros2_medkit_gateway/src/http/handlers/config_handlers.cpp [hunk 12]
```diff
@@ -345,6 +346,7 @@ void ConfigHandlers::handle_list_configurations(const httplib::Request & req, ht
 
     json response;
     response["items"] = items;
+    merge_peer_items(ctx_.aggregation_manager(), req, response, ext);
     response["x-medkit"] = ext.build();
     HandlerContext::send_json(res, response);
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/config_handlers.cpp:353]: Fixed in 72f34308 - added merge_peer_items to the empty-nodes early return path. Cross-gateway functions with no local ROS-bound nodes now get peer configurations via fan-out.

File: src/ros2_medkit_integration_tests/ros2_medkit_test_utils/constants.py [hunk 13]
```diff
@@ -40,18 +40,14 @@ def get_test_domain_id(offset=0):
...
+    mapping into the 230-232 range (DDS max domain is 232). Each
+    test's unique primary ensures a deterministic secondary. With 3
+    buckets and currently 2 multi-gateway tests, collisions are avoided.
     """
     if offset == 0:
         return DEFAULT_DOMAIN_ID
-    # Use high domain IDs (230-232) for secondary domains.
-    # offset=1 -> 230, offset=2 -> 231, offset=3 -> 232
-    return 229 + offset
+    return 230 + ((DEFAULT_DOMAIN_ID - 140) % 3)
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/ros2_medkit_test_utils/constants.py:52]: Fixed in 72f34308 - docstring now clarifies that offset is boolean (0 vs non-zero), not a multiplier.

File: src/ros2_medkit_gateway/design/aggregation.rst [hunk 14]
```diff
@@ -285,11 +285,15 @@ maps to a peer, the request is forwarded transparently:
...
-``x-medkit.failed_peers``.
+**Per-entity resource collections** (data, operations, faults, configurations,
+logs) and the global ``GET /api/v1/faults`` endpoint use real-time **fan-out**
+via ``fan_out_get()`` (handlers call the ``merge_peer_items()`` helper from
+``fan_out_helpers.hpp``): the primary gateway sends the same request to all
+healthy peers, collects the responses, and merges the ``items`` arrays. If some
+peers fail, the response body includes ``x-medkit.partial: true`` and
+``x-medkit.failed_peers``. Fan-out requests include an
+``X-Medkit-No-Fan-Out`` header to prevent recursive loops when peers have
+bidirectional aggregation.
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/design/aggregation.rst:297]: Fixed in 72f34308 - added Cross-Gateway Resource Aggregation section to rest.rst documenting fan-out behavior, x-medkit.partial and x-medkit.failed_peers fields.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/aggregation/peer_client.hpp [hunk 15]
```diff
@@ -116,7 +116,8 @@ class PeerClient {
    * @return Parsed JSON on success, error message on failure
    */
   tl::expected<nlohmann::json, std::string> forward_and_get_json(const std::string & method, const std::string & path,
-                                                                 const std::string & auth_header = "");
+                                                                 const std::string & auth_header = "",
+                                                                 const httplib::Headers & extra_headers = {});
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/aggregation/peer_client.hpp:122]: Fixed in 05b8b1fe - added @param extra_headers documentation.

File: src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py [hunk 16]
```diff
@@ -0,0 +1,651 @@
...
+      - lidar_sensor
+"""
+
+
+def _write_manifest(content):
+    """Write manifest YAML to a temporary file, return its path."""
+    fd, path = tempfile.mkstemp(suffix='.yaml', prefix='test_manifest_')
+    with os.fdopen(fd, 'w') as f:
+        f.write(content)
+    return path
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py:164]: Acknowledged - tempfile cleanup is consistent with the existing test_peer_aggregation.test.py pattern. /tmp cleanup deferred.

File: src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+  req.params.emplace("a=b", "val");
+  auto result = build_fan_out_path(req);
+  EXPECT_TRUE(result.find("a%3Db=val") != std::string::npos);
+}
+
+// =============================================================================
+// merge_peer_items
+// =============================================================================
+
+TEST(FanOutHelpers, merge_peer_items_null_aggregation_manager_is_noop) {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp:107]: Added in 05b8b1fe - 3 new tests with real AggregationManager + mock httplib server: happy-path merge, partial-failure (x-medkit.partial + failed_peers), and items-creation when result has no items key.

File: docs/api/rest.rst [hunk 18]
```diff
@@ -2213,6 +2213,30 @@ Other extensions beyond SOVD:
...
+**Cross-Gateway Resource Aggregation:**
+
+When aggregation is enabled, per-entity resource collection endpoints perform
+real-time fan-out to peer gateways. The affected endpoints are: data,
+operations, faults, configurations, logs, and the global ``GET /api/v1/faults``
+endpoint. The gateway sends the same request to all healthy peers, merges their
+``items`` arrays into the local response, and returns the combined result.
+
+If some peers are unreachable during fan-out, the response includes vendor
+metadata indicating partial results:
```
[the remark this answers is not in the document]
Comment (bburda) [reply docs/api/rest.rst]: Fixed in fe22a079.

File: docs/tutorials/multi-instance.rst [hunk 19]
```diff
@@ -224,9 +224,10 @@ When a peer goes down, the gateway handles it gracefully:
    Remote entities from the unhealthy peer are dropped from the cache on the
    next refresh cycle.
 
-3. **Fault fan-out returns partial results**: ``GET /api/v1/faults`` performs
-   real-time fan-out and includes ``x-medkit.partial: true`` and
-   ``x-medkit.failed_peers`` in the response when some peers are unreachable.
+3. **Resource fan-out returns partial results**: Per-entity resource collection
+   endpoints (data, operations, faults, configurations, logs) perform real-time
+   fan-out. When some peers are unreachable, the response includes
```
[the remark this answers is not in the document]
Comment (bburda) [reply docs/tutorials/multi-instance.rst]: Fixed in fe22a079.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 20]
```diff
@@ -547,6 +550,8 @@ void FaultHandlers::handle_list_faults(const httplib::Request & req, httplib::Re
         ext.add("clusters", result.data["clusters"]);
       }
 
+      merge_peer_items(ctx_.aggregation_manager(), req, response, ext);
+      ext.add("count", response["items"].size());
       response["x-medkit"] = ext.build();
       HandlerContext::send_json(res, response);
     } else {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp:557]: Fixed in fe22a079 - renamed to local_muted_count and local_cluster_count to make clear these are local-only FaultManager metrics, not aggregated from peers.

File: docs/config/aggregation.rst [hunk 21]
```diff
@@ -366,16 +366,25 @@ during periodic cache refresh cycles. These endpoints do not perform real-time
...
+- ``GET /api/v1/{entity_type}/{id}/operations``
+- ``GET /api/v1/{entity_type}/{id}/faults``
+- ``GET /api/v1/{entity_type}/{id}/configurations``
+- ``GET /api/v1/{entity_type}/{id}/logs``
+- ``GET /api/v1/faults`` (global fault list)
+
+If a peer is unreachable during fan-out, the response body includes:
 
 - ``x-medkit.partial: true`` in the JSON response body
 - ``x-medkit.failed_peers`` listing which peers failed
```
[the remark this answers is not in the document]
Comment (bburda) [reply docs/config/aggregation.rst]: Fixed in fe22a079.

File: src/ros2_medkit_gateway/src/aggregation/aggregation_manager.cpp [hunk 22]
```diff
@@ -559,7 +559,7 @@ AggregationManager::FanOutResult AggregationManager::fan_out_get(const std::stri
       PeerResult pr;
       pr.peer_name = peer->name();
 
-      auto result = peer->forward_and_get_json("GET", path, auth_header);
+      auto result = peer->forward_and_get_json("GET", path, auth_header, {{"X-Medkit-No-Fan-Out", "1"}});
       if (!result.has_value()) {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/aggregation/aggregation_manager.cpp:563]: Fixed in fe22a079 - forward_request now propagates X-Medkit-No-Fan-Out header to prevent recursive loops in bidirectional peering.

File: src/ros2_medkit_gateway/src/http/handlers/health_handlers.cpp [hunk 23]
```diff
@@ -23,6 +23,7 @@
 #include "ros2_medkit_gateway/discovery/discovery_manager.hpp"
 #include "ros2_medkit_gateway/gateway_node.hpp"
 #include "ros2_medkit_gateway/http/error_codes.hpp"
+#include "ros2_medkit_gateway/http/fan_out_helpers.hpp"
 #include "ros2_medkit_gateway/http/http_utils.hpp"
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/health_handlers.cpp:27]: Fixed in 2b899258 - added explicit x_medkit.hpp include.

File: src/ros2_medkit_gateway/src/http/handlers/log_handlers.cpp [hunk 24]
```diff
@@ -21,7 +21,7 @@
 
 #include "ros2_medkit_gateway/gateway_node.hpp"
 #include "ros2_medkit_gateway/http/error_codes.hpp"
-#include "ros2_medkit_gateway/http/x_medkit.hpp"
+#include "ros2_medkit_gateway/http/fan_out_helpers.hpp"
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/log_handlers.cpp:24]: Fixed in 2b899258 - added explicit x_medkit.hpp include.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 25]
```diff
@@ -0,0 +1,85 @@
...
+
+inline std::string url_encode_param(const std::string & value) {
+  std::string result;
+  result.reserve(value.size());
+  static constexpr char kHex[] = "0123456789ABCDEF";
+  for (const auto ch : value) {
+    auto c = static_cast<unsigned char>(ch);
+    if (std::isalnum(c) != 0 || c == '-' || c == '_' || c == '.' || c == '~') {
+      result += static_cast<char>(c);
+    } else {
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:39]: Fixed in 2b899258 - replaced std::isalnum with explicit ASCII range checks (A-Z, a-z, 0-9) to be locale-independent.

File: src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py [hunk 26]
```diff
@@ -0,0 +1,651 @@
...
+
+"""Integration tests for cross-ECU resource aggregation via peer fan-out.
+
+Launches TWO hybrid-mode gateways in separate DDS domains with a shared
+cross-ECU function 'vehicle_health' defined in manifests:
+
+- Primary (port 0, domain 0): aggregation enabled, manages powertrain nodes
+  (temp_sensor, rpm_sensor, calibration service).
+- Secondary (port 1, domain 1): standalone, manages chassis + perception nodes
+  (pressure_sensor, actuator, lidar_sensor with faulty params).
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_integration_tests/test/features/test_cross_ecu_fanout.test.py]: Fixed in 49fcaf0e - changed to 'port offset 0/1' and 'domain offset 0/1'.

File: src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp [hunk 27]
```diff
@@ -0,0 +1,316 @@
...
+  // Don't check health - peer starts as unhealthy, fan_out_get skips it.
+  // Force healthy flag via a mock health endpoint that 404s:
+  // Actually, fan_out_get only queries healthy peers. With no health check,
+  // peers default to unhealthy and fan_out returns empty. Instead, use a
+  // mock that returns health OK but fails on the actual request.
+  MockServer mock;
+  mock.server().Get("/api/v1/health", [](const httplib::Request &, httplib::Response & res) {
+    res.set_content(R"({"status":"healthy"})", "application/json");
+  });
+  // No handler for /api/v1/functions/f1/logs -> peer returns 404
```
[the remark this answers is not in the document]
Comment (bburda) [reply src/ros2_medkit_gateway/test/test_fan_out_helpers.cpp]: Fixed in 49fcaf0e - removed dead code (unused config/agg block).

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp [hunk 28]
```diff
@@ -547,6 +548,12 @@ void FaultHandlers::handle_list_faults(const httplib::Request & req, httplib::Re
         ext.add("clusters", result.data["clusters"]);
       }
 
+      merge_peer_items(ctx_.aggregation_manager(), req, response, ext);
+      ext.add("count", response["items"].size());
+      // muted_count and cluster_count are local-only metrics from FaultManager.
+      // They are not aggregated from peers (peer correlation data is opaque).
+      ext.add("local_muted_count", result.data["muted_count"]);
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: **[major]** Breaking API change: `muted_count` and `cluster_count` renamed to `local_muted_count` and `local_cluster_count`. This is unconditional - runs even when aggregation is disabled (`agg == nullptr` short-circuits `merge_peer_items` but the field rename here happens regardless). Clients parsing `x-medkit.muted_count` (Web UI FaultsPanel, MCP `sovd_list_faults`) will break.

Suggestion: either preserve old field names alongside new ones for one release cycle, or ship a Web UI update in the same PR, or document this as a BREAKING CHANGE in the changelog.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+    return;
+  }
+  auto fan_path = build_fan_out_path(req);
+  auto fan_result = agg->fan_out_get(fan_path, req.get_header_value("Authorization"));
+  if (fan_result.merged_items.is_array() && !fan_result.merged_items.empty()) {
+    if (!result.contains("items") || !result["items"].is_array()) {
+      result["items"] = nlohmann::json::array();
+    }
+    for (const auto & item : fan_result.merged_items) {
+      result["items"].push_back(item);
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp]: **[minor]** No deduplication of items across peers. If two peers return items with the same `id` (possible for shared function entities or broadcast fault scenarios), duplicates appear in the response. Also no `is_object()` type check on individual items - a peer returning `"items": [42, null]` would push non-object values into the merged array.

Suggestion: add `if (!item.is_object()) continue;` guard here, and consider dedup by `id` field: `if (seen_ids.count(item.value("id", ""))) continue;`. Or document that dedup is the client's responsibility.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+inline void merge_peer_items(AggregationManager * agg, const httplib::Request & req, nlohmann::json & result,
+                             XMedkit & ext) {
+  if (agg == nullptr) {
+    return;
+  }
+  if (req.has_header("X-Medkit-No-Fan-Out")) {
+    return;
+  }
+  auto fan_path = build_fan_out_path(req);
+  auto fan_result = agg->fan_out_get(fan_path, req.get_header_value("Authorization"));
```
Comment (mfaferek93) [inline src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:80]: **[minor]** Fan-out is synchronous and blocking on the HTTP handler thread. With the default 2000ms peer timeout, a single slow/dead peer holds the handler for up to 2s (parallel via std::async so it's max not sum across peers). Under concurrent load during a peer outage, this could exhaust httplib's thread pool.

Not blocking for this PR (existing `fan_out_get` for fleet-wide faults has the same characteristic), but worth documenting the thread-pool interaction and considering a shorter default timeout for per-entity fan-out vs fleet-wide.

Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:80]: Acknowledged - same constraint as existing fan_out_get for global /faults. Not blocking for this PR.

File: src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp (hunk 28 above)
Comment (bburda) [reply src/ros2_medkit_gateway/src/http/handlers/fault_handlers.cpp]: Fixed in 1825625e - reverted to original muted_count/cluster_count field names. These are always local FaultManager metrics regardless of aggregation state.

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp (hunk 29 above)
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp]: Fixed in 1825625e - added is_object() guard to skip non-object items from peer responses. Deduplication deferred (would require defining identity semantics per resource type).

File: src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp (hunk 30 above)
Comment (bburda) [reply src/ros2_medkit_gateway/include/ros2_medkit_gateway/http/fan_out_helpers.hpp:80]: Fixed in 215408d7 - merge_peer_items now checks healthy_peer_count() == 0 and returns immediately, avoiding blocking handler threads when all peers are down. Thread-pool interaction and timeout implications documented in design doc with warning admonition.

**Review by mfaferek93 — APPROVED** [review]
